File: chime/src/main/scala/com/arkr/chime/pack/Pack.scala
package com.arkr.chime.pack

import com.arkr.chime.cmd.Cmd
import io.netty.buffer.ByteBuf

/**
  * User: tanhuayou  
  * Date: 2018/7/4
  */
class Pack(val cmd: Int, val data: Array[Byte]) {

  def encode(out: ByteBuf): Unit = {
    val sumLen = 4 + 1 + 4 + data.length
    out.writeInt(sumLen)
    out.writeByte(Pack.version)
    out.writeInt(cmd)
    out.writeBytes(data)
  }

}

object Pack {
  private val version = 1

  def apply(cmd: Int, data: Array[Byte]): Pack = new Pack(cmd, data)

  def decode(in: ByteBuf): Pack = {
    val oldReadIdx = in.readerIndex()

    val allBytes = in.readableBytes()
    if (allBytes >= 4) {
      val sumLength = in.readInt()
      if (allBytes == sumLength) {
        val versionByte = in.readByte()
        if (versionByte == version) {
          val cmd = in.readUnsignedInt().intValue()
          if (Cmd.exist(cmd)) {
            val receive = new Array[Byte](allBytes - 4 - 1 - 4)
            in.readBytes(receive)
            return apply(cmd, receive)
          }
        }
      }
    }

    in.readerIndex(oldReadIdx)
    null
  }

}

File: framework/netty/in-action/src/main/java/com/in/action/codec/Pack.java
package com.in.action.codec;

import io.netty.buffer.ByteBuf;

/**
 * User: tanhuayou
 * Date: 2018/7/4
 */
public class Pack {
    private static final byte VERSION = 1;

    private short cmd;
    private byte[] body;

    /**
     * 4+1+2+x
     * LEN+VERSION+CMD+BODY
     */
    public void encode(ByteBuf out) {
        out.writeInt(7 + body.length);
        out.writeByte(VERSION);
        out.writeShort(cmd);
        out.writeBytes(body);
    }

    public static Pack decode(ByteBuf in) {
        int oldReadIndex = in.readerIndex();

        int readableBytes = in.readableBytes();
        int length = in.readInt();
        if (length == readableBytes) {
            byte version = in.readByte();
            if (version == VERSION) {
                short cmd = in.readShort();
                if (cmd >= 0) {
                    byte[] bytes = new byte[readableBytes - 7];
                    in.readBytes(bytes);
                    Pack pack = new Pack();
                    pack.cmd = cmd;
                    pack.body = bytes;
                    return pack;
                }
            }
        }

        in.readerIndex(oldReadIndex);
        return null;
    }

    public short getCmd() {
        return cmd;
    }

    public Pack setCmd(short cmd) {
        this.cmd = cmd;
        return this;
    }

    public byte[] getBody() {
        return body;
    }

    public Pack setBody(byte[] body) {
        this.body = body;
        return this;
    }
}
